### PR TITLE
[SYCL] Limit allocation size in atomic_memory_order_seq_cst test to a safe maximum size

### PR DESCRIPTION
**Pull request by GeorgeWeb**
The `atomic_memory_order_seq_cst.cpp` test only checks `info::device::max_mem_alloc_size` to limit the maximum allocation. It does not check the actual available on device. This lead to allocating exceeding amount of memory that I caught while running the test on Nvidia GTX 1650 sm_75, because the test does not check the amount of memory on the device that is free according to the OS. This commit makes it consider `ext::intel::info::device::free_memory` when selecting the number of iterations to perform. While that is still an _intel_ extension, it is currently implemented on the relevant backend targets for this test.

### CODE REVIEW

**Comment by GeorgeWeb** [comment]
@sergey-semenov Hi, is it possible to have a look at this. I've forgotten it's been up for while. It's not a prio, just that the test fails on some (Nvidia) GPUs.
The result of `info::device::max_mem_alloc_size` is not guaranteed to be exactly available on the physical hardware at the time of the allocation request, hence checking `info::device::free_memory` which accurately returns the currently available size in bytes free for allocation.

File: sycl/test-e2e/AtomicRef/atomic_memory_order_seq_cst.cpp [hunk 1]
```diff
@@ -5,13 +5,23 @@
...
-      device.get_info<info::device::max_mem_alloc_size>() / sizeof(char);
+  uint64_t max_alloc_size = device.get_info<info::device::max_mem_alloc_size>();
+  // If querying free memory is supported, use that as the max for allocation.
+  if (device.has(aspect::ext_intel_free_memory)) {
+    namespace intel = sycl::ext::intel;
+    uint64_t free_memory = device.get_info<intel::info::device::free_memory>();
+    max_alloc_size = std::min(max_alloc_size, free_memory);
+  } else {
+    std::cout << "Warning: free_memory device info query not supported. "
+              << "Befawre of allocating too much memory on the device.\n";
```
Comment (steffenlarsen) [inline sycl/test-e2e/AtomicRef/atomic_memory_order_seq_cst.cpp:22]:
```suggestion
              << "Beware of allocating too much memory on the device.\n";
```

File: sycl/test-e2e/AtomicRef/atomic_memory_order_seq_cst.cpp [hunk 2]
```diff
@@ -5,13 +5,23 @@
...
 
 constexpr size_t N_items = 128;
 
 size_t CalculateIterations(device &device, size_t iter_cap) {
-  uint64_t max_chars_alloc =
-      device.get_info<info::device::max_mem_alloc_size>() / sizeof(char);
+  uint64_t max_alloc_size = device.get_info<info::device::max_mem_alloc_size>();
+  // If querying free memory is supported, use that as the max for allocation.
+  if (device.has(aspect::ext_intel_free_memory)) {
+    namespace intel = sycl::ext::intel;
```
Comment (steffenlarsen) [inline sycl/test-e2e/AtomicRef/atomic_memory_order_seq_cst.cpp:17]: Nit: Seems a little overkill to declare a namespace alias for one use.

**Review by steffenlarsen — APPROVED** [review]
Overall LGTM! Just some small nits.

File: sycl/test-e2e/AtomicRef/atomic_memory_order_seq_cst.cpp (hunk 2 above)
Comment (GeorgeWeb) [reply sycl/test-e2e/AtomicRef/atomic_memory_order_seq_cst.cpp:17]: Yeah, thought it made the variable assignment here a readable one-liner this way, but it is overkill I guess. I'll push  the updates, cheers!

Comment (GeorgeWeb) [reply sycl/test-e2e/AtomicRef/atomic_memory_order_seq_cst.cpp]: Done.

**Review by steffenlarsen — APPROVED** [review]
🐙

**Comment by GeorgeWeb** [comment]
@intel/llvm-gatekeepers This one looks good to merge now. Thanks!

**Pull request merged**